v2/pkg/assetserver/assetserver_webview.go

```diff
@@ -0,0 +1,162 @@
+		}
+
+		dispatchC := make(chan webview.Request)
+		go queueingDispatcher(dispatchC, workerC)
+
+		d.dispatchReqC = dispatchC
+	})
+
+	if err := req.AddRef(); err != nil {
+		uri, _ := req.URL()
+		d.logError("Unable to call AddRef for request '%s'", uri)
+		return
+	}
+
+	d.dispatchReqC <- req
+}
+
+// processHTTPRequest processes the HTTP Request by faking a golang HTTP Server.
+// The request will be finished with a StatusNotImplemented code if no handler has written to the response.
+func (d *AssetServer) processWebViewRequest(r webview.Request) {
+	wrw := r.Response()
+	defer wrw.Finish()
+
+	var rw http.ResponseWriter = &contentTypeSniffer{rw: wrw} // Make sure we have a Content-Type sniffer
+	defer rw.WriteHeader(http.StatusNotImplemented)           // This is a NOP when a handler has already written and set the status
+
+	uri, err := r.URL()
+	if err != nil {
+		d.logError("Error processing request, unable to get URL: %s (HttpResponse=500)", err)
+		http.Error(rw, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	method, err := r.Method()
+	if err != nil {
+		d.webviewRequestErrorHandler(uri, rw, fmt.Errorf("HTTP-Method: %w", err))
+		return
+	}
+
+	header, err := r.Header()
+	if err != nil {
+		d.webviewRequestErrorHandler(uri, rw, fmt.Errorf("HTTP-Header: %w", err))
+		return
+	}
+
+	body, err := r.Body()
+	if err != nil {
+		d.webviewRequestErrorHandler(uri, rw, fmt.Errorf("HTTP-Body: %w", err))
+		return
+	}
+
+	if body == nil {
+		body = http.NoBody
+	}
+	defer body.Close()
+
+	req, err := http.NewRequest(method, uri, body)
+	if err != nil {
+		d.webviewRequestErrorHandler(uri, rw, fmt.Errorf("HTTP-Request: %w", err))
+		return
+	}
+	req.Header = header
+
+	if req.RemoteAddr == "" {
+		// 192.0.2.0/24 is "TEST-NET" in RFC 5737
+		req.RemoteAddr = "192.0.2.1:1234"
+	}
+
+	if req.RequestURI == "" && req.URL != nil {
+		req.RequestURI = req.URL.String()
+	}
+
+	if req.ContentLength == 0 {
+		req.ContentLength, _ = strconv.ParseInt(req.Header.Get(HeaderContentLength), 10, 64)
+	} else {
+		req.Header.Set(HeaderContentLength, fmt.Sprintf("%d", req.ContentLength))
+	}
+
+	if host := req.Header.Get(HeaderHost); host != "" {
+		req.Host = host
+	}
+
+	if expectedHost := d.ExpectedWebViewHost; expectedHost != "" && expectedHost != req.Host {
+		d.webviewRequestErrorHandler(uri, rw, fmt.Errorf("expected host '%s' in request, but was '%s'", expectedHost, req.Host))
+		return
+	}
+
+	d.ServeHTTP(rw, req)
+}
+
+func (d *AssetServer) webviewRequestErrorHandler(uri string, rw http.ResponseWriter, err error) {
+	logInfo := uri
+	if uri, err := url.ParseRequestURI(uri); err == nil {
+		logInfo = strings.Replace(logInfo, fmt.Sprintf("%s://%s", uri.Scheme, uri.Host), "", 1)
+	}
+
+	d.logError("Error processing request '%s': %s (HttpResponse=500)", logInfo, err)
+	http.Error(rw, err.Error(), http.StatusInternalServerError)
+}
+
+func queueingDispatcher(in <-chan webview.Request, out chan<- webview.Request) {
+	q := []webview.Request{}
```

Comment:
Yeah, that would probably good. Just using a pre-allocated slice would probably still be not very optimal, since it's only ever appended to the end and subsliced from the beginning. So after 100 requests it would start allocating again.

I think we should switch to some optimized FIFO queue, do you know a good go lib?

Reply:
It now uses a ringqueue buffer with minimal size of 50.